From f557cfb1329272fdfc854a7388e175c95fe7f7ed Mon Sep 17 00:00:00 2001
From: Robin Gong <yibin.gong@nxp.com>
Date: Tue, 6 Jun 2017 16:56:49 +0800
Subject: [PATCH 1870/5242] MLK-15014 dma: fsl-edma-v3: clear DONE before E_SG
 enabled

commit  f2d83ee89b6549c60caac5892b25f084dda51957 from
https://source.codeaurora.org/external/imx/linux-imx.git

Below described in RM, otherwise, channel error status(CHa_ES)
may be triggered:
The user must clear the CHa_CSR[DONE] bit before writing the
TCDa_CSR[MAJORELINK] or TCDa_CSR[ESG] bits.

Signed-off-by: Robin Gong <yibin.gong@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/fsl-edma-v3.c |    5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/dma/fsl-edma-v3.c b/drivers/dma/fsl-edma-v3.c
index c20306c..f782849 100644
--- a/drivers/dma/fsl-edma-v3.c
+++ b/drivers/dma/fsl-edma-v3.c
@@ -421,6 +421,11 @@ static void fsl_edma3_set_tcd_regs(struct fsl_edma3_chan *fsl_chan,
 
 	writel(le32_to_cpu(tcd->dlast_sga), addr + EDMA_TCD_DLAST_SGA);
 
+	/* Must clear CHa_CSR[DONE] bit before enable TCDa_CSR[ESG] */
+	if ((EDMA_TCD_CSR_E_SG | le16_to_cpu(tcd->csr)) &&
+		EDMA_CH_CSR_DONE | readl(addr + EDMA_CH_CSR))
+		writel(EDMA_CH_CSR_DONE, addr + EDMA_CH_CSR);
+
 	writew(le16_to_cpu(tcd->csr), addr + EDMA_TCD_CSR);
 }
 
-- 
1.7.9.5

